src/common/numbers/integer.cpp

```diff
@@ -0,0 +1,424 @@
+        auto digit = valueAndAdvance<char>(&it) - '0';
+        result += digit * std::pow(10, i);
+      }
+    }
+    this->slices.push_back(result);
+  }
+}
+
+Integer::Integer(intmax_t value) : _positive(value >= 0) {
+  value = std::abs(value);
+  while (value > 0) {
+    auto slice = value % static_cast<value_t>(SLICE_SIZE);
+    slices.push_back(static_cast<uintmax_t>(slice));
+    value = (value - slice) / SLICE_SIZE;
+  }
+  slices.shrink_to_fit();
+}
+
+std::string Integer::toString() const {
+  if (slices.empty()) {
+    ensure(_positive);
+    return "0";
+  }
+
+  std::string result;
+  for (auto slice : slices) {
+    auto nextSlice = std::to_string(slice);
+    auto paddedSlice = Puzzles::padLeading(nextSlice, 9, '0');
+    result.insert(0, paddedSlice);
+  }
+
+  result = Puzzles::trimLeading(result, '0');
+
+  if (!_positive) {
+    result.insert(0, 1, '-');
+  }
+
+  return result;
+}
+
+inline compat::strong_ordering compareSlices(const std::vector<Integer::value_t> &left,
+                                             const std::vector<Integer::value_t> &right) {
+  auto lengthComparison = compat::compare(left.size(), right.size());
+  if (lengthComparison != compat::strong_ordering::equal) {
+    return lengthComparison;
+  }
+
+  auto lit = left.crbegin(), rit = right.crbegin();
+  while (lit != left.crend()) {
+    ensure(rit != right.crend());
+
+    auto sliceComparison = compat::compare(*lit, *rit);
+    if (sliceComparison == compat::strong_ordering::equal) {
+      ++lit;
+      ++rit;
+      continue;
+    }
+
+    return sliceComparison;
+  }
+
+  return compat::strong_ordering ::equal;
+}
+
+Integer Integer::operator+(const Integer &o) const {
+  if (slices.empty()) return o;
+  if (o.slices.empty()) return *this;
+
+  auto sameSign = this->positive() == o.positive();
+  if (!sameSign && this->slices == o.slices) {
+    return Integer(0);
+  }
+
+  bool finalSign;
+  typeof(slices.cbegin()) lit, rit;
+  typeof(slices.cend()) lend, rend;
+
+  // if sameSign, we don't care about the order, so we can avoid the compareSlices call
+  if (sameSign || compareSlices(this->slices, o.slices) != compat::strong_ordering::less) {
+    finalSign = this->_positive;
+    lit = this->slices.cbegin();
+    lend = this->slices.cend();
+
+    rit = o.slices.cbegin();
+    rend = o.slices.cend();
+  } else {
+    finalSign = o._positive;
+    lit = o.slices.cbegin();
+    lend = o.slices.cend();
+
+    rit = this->slices.cbegin();
+    rend = this->slices.cend();
+  }
+
+  std::vector<value_t> result;
+  result.reserve(std::max(this->slices.size(), o.slices.size()) + 1);
+
+  int_fast64_t carryOver = 0;
+
+  while (lit != lend || rit != rend) {
+    int_fast64_t lbit = lit == lend ? 0 : valueAndAdvance<value_t>(&lit);
+    int_fast64_t rbit = rit == rend ? 0 : valueAndAdvance<value_t>(&rit);
+    int_fast64_t sum = sameSign ? lbit + rbit + carryOver : lbit - rbit - carryOver;
+
+    ensure(sum >= -1000000000 && sum <= 1999999999);
```

Comment:
Use the new constants instead
```suggestion
    ensure(sum >= (SLICE_SIZE * static_cast<intmax_t>(-1)) && sum <= (SLICE_SIZE * 2 - 1));
```